pkgs/top-level/all-packages.nix

```diff
@@ -2664,7 +2664,21 @@ with pkgs;
     inherit (darwin.apple_sdk_11_0.frameworks) AudioUnit Carbon Cocoa;
   };
 
-  duckstation = qt6Packages.callPackage ../applications/emulators/duckstation { };
+  duckstation = callPackage ../by-name/du/duckstation/package.nix {
+    shaderc = shaderc.overrideAttrs (old: {
+      patches = (old.patches or [ ]) ++ [
+        (fetchpatch {
+          url = "file://${duckstation.src}/scripts/shaderc-changes.patch";
```

Comment:
This is a bit weird. Is there an IFD limitation to use this:

```nix
patches = [
  ${duckstation.src}/scripts/shaderc-changes.patch
];
```

? I was also thinking, since you promote the `pkgs/by-name` usage, how about putting these details inside the expression, inside the `let` `in`? Then you could leave `all-packages.nix` completely.

Reply:
I am not aware of IFD limitations.
I used this fetchpatch because 
1. fetchpatch is preferrred to vendoring
2. the patch can't be applied without the exclusions listed in the code
3. fetchpatch provides a clean and understandable interface to patch manipulation

i will answer the second part later.

Reply:
> I was also thinking, since you promote the `pkgs/by-name` usage, how about putting these details inside the expression, inside the `let` `in`? Then you could leave `all-packages.nix` completely.

The current documentation about `by-name` puts some effort in preserving the `.override` interface:

  - https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults
  - https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#manual-migration-guidelines

That being said, we have two possibilities:

1. Expect an unpatched `shaderc`, patch and compile it inside the expression, and after that patch duckstation itself
2. Expect an already patched `shaderc` and use it as is.

The second approach looks more natural, since it does not subject the user to a bad surprise.

If by any reason the user of `duckstation` expression wants to override `shaderc`, theey is free to do so.
In the first case, the user would be forced to undo the expression, probably using `overrideAttrs`.

In a certain sense it is an application of [POLA](https://docs.freebsd.org/en/books/handbook/glossary/#pola-glossary).

Reply:
Hmm it is not clear from what you are saying, why does `shaderc` needs to get this patch.

- If without the patch `duckstation` is not usable, then I don't think it is too bad if it is not trivial to override the `shaderc` that is eventually used.
- Eitherway, when a user is attempting to override a package, usually they need to read the source code anyway, so I don't think the POLA argument applies here, because if anyway you have to read the source code you won't be "astonished".
- You can still provide the user the ability to override the `shaderc` that is actually used in the package, by defining it in the `passthru`, and use it as so:

```nix
stdenv.mkDerivation(finalAttrs: {
  # ...
  buildInputs = [
    # ...
    finalAttrs.passthru.shaderc
    # ...
  ];
  # ...
  passthru = {
    shaderc = shaderc.overrideAttrs(old: {
      # ...
    });
  };
  # ...
}
```

This way, the user will be able to override it by overriding the `passthru`. Also, if they will use `patches = old.patches + [ (their patches) ];`, they should retain the patches provided by `duckstation`, and the overriding experience will be pleasant IMO.

Reply:
Without the patch, Duckstation fail to compile.

Reply:
It has an extra problem: if the shaderc is patched and built inside duckstation, then duckstation requires a nested building, and an expression should build only one thing, not two.

Reply:
> Hmm it is not clear from what you are saying, why does `shaderc` needs to get this patch.

Basically it is a vendoring, but made via patches instead of using git modules or scripts.

> 
>     * If without the patch `duckstation` is not usable, then I don't think it is too bad if it is not trivial to override the `shaderc` that is eventually used.

It has an extra problem: if the shaderc is patched and built inside duckstation, then duckstation requires a nested building, and an expression should build only one thing, not two.

> 
>     * Eitherway, when a user is attempting to override a package, usually they need to read the source code anyway, so I don't think the POLA argument applies here, because if anyway you have to read the source code you won't be "astonished".

The expression should be encapsulated, acting like a blackbox. For all effects and purposes, Nixpkgs is an API,  sorta. 
It is not transparent to declare `shaderc` as an input when it is an implementation detail of that expression.

> 
>     * You can still provide the user the ability to override the `shaderc` that is actually used in the package, by defining it in the `passthru`, and use it as so:

This is too confusing.

First, usually passthru is used to emit informations about the expression, not as an input.
Second, your suggestion looks like a convoluted form of doing the same: patching an input.

Reply:
> > Hmm it is not clear from what you are saying, why does `shaderc` needs to get this patch.
> 
> Basically it is a vendoring, but made via patches instead of using git modules or scripts.

This is not an answer to the question. @SuperSamus' answered this.

> > If without the patch `duckstation` is not usable, then I don't think it is too bad if it is not trivial to override the `shaderc` that is eventually used.
> 
> It has an extra problem: if the shaderc is patched and built inside duckstation, then duckstation requires a nested building, and an expression should build only one thing, not two.

I didn't mean to suggest to not patch `shaderc` at all, or to build it inside `duckstation`, I was only talking about how to define, and in what scope, the patched `shaderc`.

> > Eitherway, when a user is attempting to override a package, usually they need to read the source code anyway, so I don't think the POLA argument applies here, because if anyway you have to read the source code you won't be "astonished".
> 
> The expression should be encapsulated, acting like a blackbox.

I disagree with that. It sounds like a philosophy you developed personally, and I never heard it before. Moreover, we had discussions at [RFC 166](https://github.com/NixOS/rfcs/pull/166) inventing and using Nix "docstring" format, to document near the inputs of an expression, exactly the kind of information about the role of inputs.

> For all effects and purposes, Nixpkgs is an API, sorta. It is not transparent to declare `shaderc` as an input when it is an implementation detail of that expression.

Sounds like this argument was derived from the derived from the blackbox philosophy, and I disagree with it as well.

> > You can still provide the user the ability to override the `shaderc` that is actually used in the package, by defining it in the `passthru`, and use it as so:
> 
> This is too confusing.

If you explain this in comments, both near the initial inputs, and near the `finalAttrs.passthru.shaderc` line in `buildInputs`. It is not confusing IMO.

> First, usually passthru is used to emit informations about the expression, not as an input.

The fact that `shaderc` is overridden, and it is not the same `shaderc` from `pkgs`, is in a way an information about the expression, that you help expose, by putting it in the `passthru`.

Secondly, this practice is very useful in cases such as this:

- `pkg1` requires a patched `pkg2`, named `pkg2'`.
- `pkg3` requires both `pkg1`, and `pkg2'`.

Defining `pkg1.passthru.pkg2 = pkg2';` helps accessing the overriden `pkg2` in the expression of `pkg3`. [This method is used in octave](https://github.com/NixOS/nixpkgs/blob/52493b0dfa0031b9b5f3edff7a96aea74d5a741b/pkgs/development/interpreters/octave/default.nix#L186-L202), and the `passthru` packages are later [used in `octavePackages`](https://github.com/NixOS/nixpkgs/blob/52493b0dfa0031b9b5f3edff7a96aea74d5a741b/pkgs/top-level/octave-packages.nix#L133-L136)

> Second, your suggestion looks like a convoluted form of doing the same: patching an input.

Correct, I am not against patching `shaderc`, I am only discussing where and how the patch should be applied.

Reply:
> This is not an answer to the question. @SuperSamus' answered this.

SuperSamus said what happens without the patch. 
I explained the origin and role of the patch.
Both are acceptable answers.

> I didn't mean to suggest to not patch `shaderc` at all, or to build it inside `duckstation`, I was only talking about how to define, and in what scope, the patched `shaderc`.

I am talking about this all along.
Is there anything confusing about what I am talking about?

> > The expression should be encapsulated, acting like a blackbox.
> 
> I disagree with that. It sounds like a philosophy you developed personally, and I never heard it before. 

You are shifting the burden here.

First, do you have a more technical reason besides your personal disagreement and ignorance? 
Personal disagreement and ignorance are not sufficient reasons to block a PR.

Second, there is nothing hindering me, or any contributor for that matter, of following philosophies developed personally while writing Nix expressions for Nixpkgs.

> Moreover, we had discussions at [RFC 166](https://github.com/NixOS/rfcs/pull/166) inventing and using Nix "docstring" format, to document near the inputs of an expression, exactly the kind of information about the role of inputs.

This is completely irrelevant to this thread.
The docstring standard is, or should be, completely capable to document any of the approaches discussed here.

> > For all effects and purposes, Nixpkgs is an API, sorta. It is not transparent to declare `shaderc` as an input when it is an implementation detail of that expression.
> 
> Sounds like this argument was derived from the derived from the blackbox philosophy, and I disagree with it as well.

Do you have a more technical reason besides your personal disagreement?

> If you explain this in comments, both near the initial inputs, and near the `finalAttrs.passthru.shaderc` line in `buildInputs`. It is not confusing IMO.

If it needs careful commenting, then it is confusing, or at least it is unexpected.

>  [This method is used in octave](https://github.com/NixOS/nixpkgs/blob/52493b0dfa0031b9b5f3edff7a96aea74d5a741b/pkgs/development/interpreters/octave/default.nix#L186-L202), and the `passthru` packages are later [used in `octavePackages`](https://github.com/NixOS/nixpkgs/blob/52493b0dfa0031b9b5f3edff7a96aea74d5a741b/pkgs/top-level/octave-packages.nix#L133-L136)

If you are to cite precedents, then I can cite occurrences similar to mine:

https://github.com/NixOS/nixpkgs/blob/52493b0dfa0031b9b5f3edff7a96aea74d5a741b/pkgs/top-level/all-packages.nix#L24296-L24312

> Correct, I am not against patching `shaderc`, I am only discussing where and how the patch should be applied.

Nice.
Do you have a more technical reason besides your personal preference?

Reply:
> The only technical consideration I will stand upon, is to not put so many details in `all-packages.nix`, and avoid editing it as much as possible (only removing lines is OK of course).

I do not like to edit `all-packages` too.
Indeed when I was doing tree-wide migration Infinisil theemself told me to stop.

However there is almost no other way to override inputs.
Many of the ideas were rejected in the [RFC 140 itself](https://github.com/NixOS/rfcs/blob/f15c1cf8a73013d32092148147abaaddbf1b5e75/rfcs/0140-simple-package-paths.md?plain=1#L357)

> I feel you prefer misunderstanding me, instead of acknowledge the subtleties I am trying to layout for you, to help you with your PR.

I am trying to see your point, but I feel that hiding the modifications over shaderc is not correct.
I will try another idea.